### PR TITLE
Fix race condition when cleaning up. 

### LINKED ISSUE #130: Fed cleanup logic is not thread safe
We are using multiple threads on the fed cleanup module, which has thread safety issues, for example:

1.  Main thread: global _check_send_thread is not None
2. _check_send_thread: clear global _check_send_thread variable to None
3. Main thread: calling _check_send_thread.join(), this would error because None object does not have join method

We should check all threading usage in ray fed and fix this.

### PR DESCRIPTION
**Pull request by jovany-wang**
In this PR, we refactor the cleanup file to solve the thread safe issue and repeated initializing issue.

Replace #142 and close #130

### CODE REVIEW

File: fed/cleanup.py [hunk 1]
```diff
@@ -23,97 +23,86 @@
...
+        def _signal_exit():
+            os.kill(os.getpid(), signal.SIGTERM)
+
+        assert self._sending_obj_refs_q is not None
+
+        while True:
+            try:
+                obj_ref = self._sending_obj_refs_q.popleft()
+            except IndexError:
+                time.sleep(0.5)
```
Comment (zhouaihui) [inline fed/cleanup.py:94]: Maybe we can reduce the sleep interval, say 0.1 seconds?

File: fed/cleanup.py [hunk 2]
```diff
@@ -23,97 +23,86 @@
...
+        thread exited, then notifys the checking thread.
+    """
+
+    def __init__(self) -> None:
+        # `deque()` is thread safe on `popleft` and `append` operations.
+        # See https://docs.python.org/3/library/collections.html#deque-objects
+        self._sending_obj_refs_q = deque()
+        self._check_send_thread = None
+        self._monitor_thread = None
+        self._lock_on_sending_q = threading.Lock()
```
Comment (zhouaihui) [inline fed/cleanup.py:49]: `_lock_on_sending_q` and `_lock_on_send_thread` are not used.

Comment (jovany-wang) [reply fed/cleanup.py:49]: good catch!

File: fed/cleanup.py (hunk 1 above)
Comment (jovany-wang) [reply fed/cleanup.py:94]: DONE

**Review by zhouaihui — APPROVED** [review]
LGTM

File: fed/cleanup.py [hunk 3]
```diff
@@ -23,97 +23,84 @@
...
+    It monitors whether the main thread is broken and it needs wait until all sending
+    objects get repsonsed.
+
+    The main logic path is:
+        A. If `fed.shutdown()` is invoked in the main thread and every thing works well,
+        the `graceful_stop()` will be invoked as well and the checking thread will be
+        notifiled to exit gracefully.
+
+        B. If the main thread are broken before sending the notification flag to the
+        sending thread, the monitor thread will detech that and it joins until the main
```
Comment (fengsp) [inline fed/cleanup.py:39]: detect

**Review by fengsp — APPROVED** [review]
LGTM

File: fed/cleanup.py (hunk 3 above)
Comment (jovany-wang) [reply fed/cleanup.py:39]: Good catch.

**Pull request merged**